conformance/tests/gateway-infrastructure.go

```diff
@@ -0,0 +1,90 @@
+package tests
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	v1 "sigs.k8s.io/gateway-api/apis/v1"
+	"sigs.k8s.io/gateway-api/conformance/utils/kubernetes"
+	"sigs.k8s.io/gateway-api/conformance/utils/suite"
+	"sigs.k8s.io/gateway-api/pkg/features"
+)
+
+func init() {
+	ConformanceTests = append(ConformanceTests, GatewayInfrastructure)
+}
+
+var GatewayInfrastructure = suite.ConformanceTest{
+	ShortName:   "GatewayInfrastructure",
+	Description: "Propagation of metadata from Gateway infrastructure to generated components",
+	Features: []features.SupportedFeature{
+		features.SupportGateway,
+		features.SupportGatewayInfrastructurePropagation,
+	},
+	Manifests: []string{
+		"tests/gateway-infrastructure.yaml",
+	},
+	Test: func(t *testing.T, s *suite.ConformanceTestSuite) {
+		ns := "gateway-conformance-infra"
+
+		kubernetes.NamespacesMustBeReady(t, s.Client, s.TimeoutConfig, []string{ns})
+
+		gwNN := types.NamespacedName{
+			Name:      "gateway-with-infrastructure-metadata",
+			Namespace: "gateway-conformance-infra",
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), s.TimeoutConfig.DefaultTestTimeout)
+		defer cancel()
+
+		t.Logf("waiting for namespace %s and Gateway %s to be ready for testing", gwNN.Namespace, gwNN.Name)
+		kubernetes.GatewayMustHaveLatestConditions(t, s.Client, s.TimeoutConfig, gwNN)
+
+		t.Logf("retrieving Gateway %s/%s", gwNN.Namespace, gwNN.Name)
+		currentGW := &v1.Gateway{}
+		err := s.Client.Get(ctx, gwNN, currentGW)
+		require.NoError(t, err, "error getting Gateway: %v", err)
+		t.Logf("verifying that the Gateway %s/%s is accepted with infrastructure declared", gwNN.Namespace, gwNN.Name)
+		kubernetes.GatewayMustHaveCondition(t, s.Client, s.TimeoutConfig, gwNN, metav1.Condition{
+			Type:   string(v1.GatewayConditionAccepted),
+			Status: metav1.ConditionTrue,
+		})
+
+		t.Logf("verifying that generated resources for Gateway %s/%s have the proper gateway name label", gwNN.Namespace, gwNN.Name)
+		// Don't check services because implementations may have special filtering logic (e.g. for LB annotations)
+		// Instead, check service accounts for the gateway-name label first and then
+		// fallback to Pod if that fails
+		annotations := make(map[string]string, len(currentGW.Spec.Infrastructure.Annotations))
+		labels := make(map[string]string, len(currentGW.Spec.Infrastructure.Labels))
+		// Need to translate from (Annotation|Label)Key to string
+		for k, v := range currentGW.Spec.Infrastructure.Annotations {
+			annotations[string(k)] = string(v)
+		}
+
+		for k, v := range currentGW.Spec.Infrastructure.Labels {
+			labels[string(k)] = string(v)
+		}
+		saList := corev1.ServiceAccountList{}
```

Comment:
This is tricky to get right. It seems like if we're going to take this approach, we'll need to:

1. Define a fairly long list of resources to check (Services, Pods, and ServiceAccounts, likely more)
1. Check all of these resources independently, using the `gateway.networking.k8s.io/gateway-name` label as part of the list criteria
1. Ensure that any matching resources had the labels specified and annotations specified. I think we'd still need to allow additional labels and annotations to be set, we'd just need to confirm that the specified labels and annotations were in the list
1. Ensure that we found at least one matching resource across the various kinds we looked at

Even with all of that, this test feels like it's testing a small subset of the scope of the GEP. I think the name of the supported feature should probably clarify that this is specifically for "In-Cluster" implementations of the feature. I don't want to rule out the idea that other implementations such as cloud-based ones could also label the underlying resources they're creating with a similar pattern.

With all of that said, I'm not the best person to evaluate the viability of this test as I'm not actively working on an in-cluster implementation of the API. Would appreciate some feedback from others on this.

/cc @arkodg @mlavacca @sunjayBhatia

Reply:
To be clear: this test only checks that at least one of those resources has the labels/annotations copied. The reason it's so lax is because I don't want to be too prescriptive about what resources implementations deploy (e.g. Deployment vs. Daemonset, etc.). I agree that the value is marginal; it's tough to write a 100% accurate conformance test for this feature, but it depends on how far we want to go

Reply:
Currently, I don't think this will work for Cilium because we don't generate either ServiceAccounts or Pods as part of our generated config - we only generate Services and (because of Cilium-specific things) Endpoints. Yes, I know they should be EndpointSlices, that's in the backlog.

I think the test will have to check Services as one option, although I agree that the only way that makes sense here is to have a list of things to check, and have the test criteria be that _at least one_ of the types has the required `gateway-name` label as well as the other labels and/or annotations. That way, if Service isn't found for an implementation, that's fine as long as ServiceAccount or Pod is.

Reply:
Thanks for chiming in! That context on Cilium's implementation is extremely helpful; I'll add checks for Services as well then 👍🏾